src/libappimage/core/PayloadIterator.cpp

```diff
@@ -0,0 +1,189 @@
+
+            bool isCompleted() { return traversal == nullptr; }
+
+            void next();
+
+            PayloadEntryType type() { return isCompleted() ? PayloadEntryType::UNKNOWN : traversal->getEntryType(); }
+
+            std::string entryName() { return isCompleted() ? std::string() : traversal->getEntryName(); }
+
+            std::string entryLink() { return isCompleted() ? std::string() : traversal->getEntryLink(); }
+
+            void extractTo(const std::string& target) {
+                // Enforce ONE PASS restriction
+                if (entryDataConsumed)
+                    throw PayloadIteratorError("Entry data consumed");
+                else
+                    entryDataConsumed = true;
+
+                if (!isCompleted()) traversal->extract(target);
+            }
+
+            std::istream& read() {
+                // Enforce ONE PASS restriction
+                if (entryDataConsumed)
+                    throw PayloadIteratorError("Entry data consumed");
+                else
+                    entryDataConsumed = true;
+
+                return isCompleted() ? emptyStream : traversal->read();
+            }
+
+            Private* beginState() { return new Private(appImage); }
+
+            Private* endState() { return new Private(appImage, true); }
+        };
+
+        PayloadIterator::PayloadIterator(const AppImage& appImage) : d(new Private(appImage)) {}
+
+        PayloadIterator::PayloadIterator(PayloadIterator&& other) noexcept { d = other.d; }
+
+        PayloadIterator& PayloadIterator::operator=(PayloadIterator&& other) noexcept {
+            d = other.d;
+            return *this;
+        }
+
+        PayloadEntryType PayloadIterator::type() { return d->type(); }
+
+        std::string PayloadIterator::path() { return d->entryName(); }
+
+        std::string PayloadIterator::link() { return d->entryLink(); }
+
+        void PayloadIterator::extractTo(const std::string& target) { d->extractTo(target); }
+
+        std::istream& PayloadIterator::read() { return d->read(); }
+
+        std::string PayloadIterator::operator*() { return d->entryName(); }
+
+        PayloadIterator& PayloadIterator::operator++() {
+            // move to the next entry in the traversal
+            d->next();
+
+            return *this;
+        }
+
+        PayloadIterator PayloadIterator::begin() { return PayloadIterator(d->beginState()); }
+
+        PayloadIterator PayloadIterator::end() { return PayloadIterator(d->endState()); }
+
+        PayloadIterator::PayloadIterator(PayloadIterator::Private* d) : d(d) {}
+
+        bool PayloadIterator::operator==(const PayloadIterator& other) const { return *d == *(other.d); }
+
+        bool PayloadIterator::operator!=(const PayloadIterator& other) const { return !(other == *this); }
+
+        /**
+         * FilesIterator Private methods
+         */
+        PayloadIterator::Private::Private(const AppImage& appImage, bool atEnd) : appImage(appImage) {
```

Comment:
What's the point of defining methods outside the class declaration for PImpl classes?

Reply:
There is not too much point on it. Amended.